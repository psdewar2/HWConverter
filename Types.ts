// Converter interfaces
export enum Unit {
  imperial = 'imperial',
  metric = 'metric',
}

export enum Label {
  // height
  meters = 'm',
  feet = 'ft',
  inches = 'in',
  // weight
  kilograms = 'kg',
  pounds = 'lbs',
}

export type ImperialMeasurements = {
  feet: number;
  inches: number;
  pounds: number;
};

export type MetricMeasurements = {
  meters: number;
  kilograms: number;
};

// Reducer
export type ScreenState = {
  imperialOn: boolean; // if false, metric is on
} & ImperialMeasurements &
  MetricMeasurements;

export enum ActionType {
  ConvertToFeetInches = 'CONVERT_TO_FT_IN',
  ConvertFeetToMeters = 'CONVERT_FT_TO_M',
  ConvertInchesToMeters = 'CONVERT_IN_TO_M',
  ConvertToKilograms = 'CONVERT_TO_KG',
  ConvertToPounds = 'CONVERT_TO_LB',
  ClearState = 'CLEAR_STATE',
  SetImperialOn = 'SET_IMPERIAL_ON',
  SaveState = 'SAVE_STATE',
}

export type ScreenAction =
  | {type: ActionType.ClearState}
  | {type: ActionType.SetImperialOn; payload: boolean}
  | {type: ActionType.SaveState; payload: ScreenState}
  | {
      type:
        | ActionType.ConvertToFeetInches
        | ActionType.ConvertFeetToMeters
        | ActionType.ConvertInchesToMeters
        | ActionType.ConvertToKilograms
        | ActionType.ConvertToPounds;
      payload: number;
    };
