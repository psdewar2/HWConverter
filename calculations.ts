export function inToM(inches: number) {
  return Number((0.0254 * inches).toFixed(2));
}

export function mToFtIn(meters: number) {
  let inches = 39.37 * meters;
  return [Math.floor(inches / 12), Number((inches % 12).toFixed(2))]; // feet, inches
}

export function lbToKg(pounds: number) {
  return Number((0.4536 * pounds).toFixed(2));
}

export function kgToLb(kilograms: number) {
  return Number((2.205 * kilograms).toFixed(2));
}
