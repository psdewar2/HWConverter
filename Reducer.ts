import {inToM, kgToLb, lbToKg, mToFtIn} from './calculations';
import {ActionType, ScreenAction, ScreenState} from './Types';

export const initialScreenState: ScreenState = {
  imperialOn: true,
  feet: 0,
  inches: 0,
  pounds: 0,
  meters: 0,
  kilograms: 0,
};

export function reducer(state: ScreenState, action: ScreenAction) {
  switch (action.type) {
    case ActionType.ClearState:
      return initialScreenState;
    case ActionType.SaveState:
      return action.payload;
    case ActionType.SetImperialOn:
      return {
        ...state,
        imperialOn: action.payload,
      };
    case ActionType.ConvertToFeetInches: {
      const meters = action.payload;
      const [feet, inches] = mToFtIn(meters);
      return {
        ...state,
        feet,
        inches,
        meters,
      };
    }
    case ActionType.ConvertToKilograms: {
      const pounds = action.payload;
      const kilograms = lbToKg(pounds);
      return {...state, pounds, kilograms};
    }
    case ActionType.ConvertFeetToMeters: {
      const feet = action.payload;
      const meters = inToM(feet * 12 + state.inches);
      return {
        ...state,
        feet,
        meters,
      };
    }
    case ActionType.ConvertInchesToMeters: {
      const inches = action.payload;
      const meters = inToM(state.feet * 12 + inches);
      return {
        ...state,
        inches,
        meters,
      };
    }
    case ActionType.ConvertToPounds: {
      const kilograms = action.payload;
      const pounds = kgToLb(kilograms);
      return {
        ...state,
        pounds,
        kilograms,
      };
    }
  }
}
