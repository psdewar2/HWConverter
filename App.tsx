/**
 * Project by Peyt Spencer Dewar
 */

import AsyncStorage from '@react-native-community/async-storage';
import React, {useEffect, useReducer} from 'react';
import {
  Button,
  Pressable,
  SafeAreaView,
  ScrollView,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import {Measurement} from './Measurement';
import {initialScreenState, reducer} from './Reducer';
import {ActionType, Label, Unit} from './Types';

export const STORAGE_KEY = 'converter';

export function App() {
  const [screenState, dispatch] = useReducer(reducer, initialScreenState);
  const {imperialOn, feet, inches, pounds, meters, kilograms} = screenState;

  useEffect(() => {
    async function getConverterState() {
      const state = await AsyncStorage.getItem(STORAGE_KEY);
      if (state) {
        dispatch({type: ActionType.SaveState, payload: JSON.parse(state)});
      }
    }

    getConverterState();
  }, []);

  return (
    <SafeAreaView>
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <View style={styles.primaryContainer}>
          <Text style={styles.primaryText}>HW Converter</Text>
          <View style={styles.buttonView}>
            <Pressable
              accessibilityLabel={'Imperial button'}
              onPress={() =>
                dispatch({type: ActionType.SetImperialOn, payload: true})
              }
              style={[
                styles.button,
                {backgroundColor: imperialOn ? 'lightgray' : 'white'},
              ]}>
              <Text
                style={{
                  fontWeight: imperialOn ? '600' : '200',
                  fontSize: 24,
                }}>
                {Unit.imperial}
              </Text>
            </Pressable>
            <Pressable
              accessibilityLabel={'Metric button'}
              onPress={() =>
                dispatch({type: ActionType.SetImperialOn, payload: false})
              }
              style={[
                styles.button,
                {backgroundColor: !imperialOn ? 'lightgray' : 'white'},
              ]}>
              <Text
                style={{
                  fontWeight: !imperialOn ? '600' : '200',
                  fontSize: 24,
                }}>
                {Unit.metric}
              </Text>
            </Pressable>
          </View>
        </View>
        <View style={styles.secondaryContainer}>
          <Text style={styles.secondaryText}>Measurements</Text>
          <View style={styles.quantityView}>
            <Text style={styles.quantityName}>Height</Text>
            <View style={styles.measurementsView}>
              {imperialOn ? (
                <>
                  <Measurement
                    label={Label.feet}
                    value={feet}
                    onChangeValue={val =>
                      dispatch({
                        type: ActionType.ConvertFeetToMeters,
                        payload: Number(val),
                      })
                    }
                  />
                  <Measurement
                    label={Label.inches}
                    value={inches}
                    onChangeValue={val =>
                      dispatch({
                        type: ActionType.ConvertInchesToMeters,
                        payload: Number(val),
                      })
                    }
                  />
                </>
              ) : (
                <Measurement
                  label={Label.meters}
                  value={meters}
                  onChangeValue={val =>
                    dispatch({
                      type: ActionType.ConvertToFeetInches,
                      payload: Number(val),
                    })
                  }
                />
              )}
            </View>
            <Text style={styles.quantityName}>Weight</Text>
            <View style={styles.measurementsView}>
              {imperialOn ? (
                <Measurement
                  label={Label.pounds}
                  value={pounds}
                  onChangeValue={val =>
                    dispatch({
                      type: ActionType.ConvertToKilograms,
                      payload: Number(val),
                    })
                  }
                />
              ) : (
                <Measurement
                  label={Label.kilograms}
                  value={kilograms}
                  onChangeValue={val =>
                    dispatch({
                      type: ActionType.ConvertToPounds,
                      payload: Number(val),
                    })
                  }
                />
              )}
            </View>
          </View>
          <View style={styles.buttonView}>
            <Button
              title={'Save'}
              onPress={() =>
                AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(screenState))
              }
            />
            <Button
              title={'Clear'}
              onPress={() => {
                dispatch({type: ActionType.ClearState});
                AsyncStorage.removeItem(STORAGE_KEY);
              }}
            />
          </View>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  primaryContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  primaryText: {
    fontSize: 40,
    fontWeight: '600',
  },
  secondaryContainer: {
    paddingHorizontal: 24,
  },
  secondaryText: {
    fontSize: 32,
    fontWeight: '600',
  },
  buttonView: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    marginVertical: 16,
  },
  button: {
    alignItems: 'center',
    borderRadius: 4,
    elevation: 3,
    justifyContent: 'center',
    paddingVertical: 12,
    paddingHorizontal: 32,
  },
  quantityName: {
    fontSize: 24,
    fontWeight: '200',
  },
  quantityView: {
    marginVertical: 16,
  },
  measurementsView: {
    flexDirection: 'row',
  },
});
