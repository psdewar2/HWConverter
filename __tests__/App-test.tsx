import React from 'react';
import {act, fireEvent, render} from '@testing-library/react-native';
import {App, STORAGE_KEY} from '../App';
import {Label, Unit} from '../Types';
import {inToM, kgToLb, lbToKg, mToFtIn} from '../calculations';
import AsyncStorage from '@react-native-community/async-storage';
import {initialScreenState} from '../Reducer';

describe('App', () => {
  // Arrange
  const component = <App />;

  it('renders without crashing', async () => {
    // Act
    const rendererJson = render(component).toJSON();
    // Assert
    expect(rendererJson).toMatchSnapshot();
  });

  it('displays imperial units', () => {
    // Arrange
    const {getByText, queryByText} = render(component);
    // Act
    fireEvent.press(getByText(Unit.imperial));
    // Assert
    expect(getByText(Label.feet)).toBeTruthy();
    expect(queryByText(Label.meters)).toBeNull();
  });

  it('displays metric units', () => {
    // Arrange
    const {getByText, queryByText} = render(component);
    // Act
    fireEvent.press(getByText(Unit.metric));
    // Assert
    expect(getByText(Label.kilograms)).toBeTruthy();
    expect(queryByText(Label.pounds)).toBeNull();
  });

  it('converts feet to meters', () => {
    // Arrange
    const feet = 6;
    const {getByDisplayValue, getByText, getAllByPlaceholderText} =
      render(component);
    // Act
    fireEvent.changeText(getAllByPlaceholderText('0')[0], feet);
    fireEvent.press(getByText(Unit.metric));
    // Assert
    expect(getByDisplayValue(inToM(feet * 12).toString())).toBeTruthy();
  });

  it('converts inches to meters', () => {
    // Arrange
    const inches = 6;
    const {getByDisplayValue, getByText, getAllByPlaceholderText} =
      render(component);
    // Act
    fireEvent.changeText(getAllByPlaceholderText('0')[1], inches);
    fireEvent.press(getByText(Unit.metric));
    // Assert
    expect(getByDisplayValue(inToM(inches).toString())).toBeTruthy();
  });

  it('converts pounds to kilograms', () => {
    // Arrange
    const pounds = 100;
    const {getByDisplayValue, getByText, getAllByPlaceholderText} =
      render(component);
    // Act
    fireEvent.changeText(getAllByPlaceholderText('0')[2], pounds);
    fireEvent.press(getByText(Unit.metric));
    // Assert
    expect(getByDisplayValue(lbToKg(pounds).toString())).toBeTruthy();
  });

  it('converts meters to feet-inches', () => {
    // Arrange
    const meters = 1.5;
    const {getByDisplayValue, getByText, getAllByPlaceholderText} =
      render(component);
    // Act
    fireEvent.press(getByText(Unit.metric));
    fireEvent.changeText(getAllByPlaceholderText('0')[0], meters);
    fireEvent.press(getByText(Unit.imperial));
    // Assert
    const [feet, inches] = mToFtIn(meters);
    expect(getByDisplayValue(feet.toString())).toBeTruthy();
    expect(getByDisplayValue(inches.toString())).toBeTruthy();
  });

  it('converts kilograms to pounds', () => {
    // Arrange
    const kilograms = 100;
    const {getByDisplayValue, getByText, getAllByPlaceholderText} =
      render(component);
    // Act
    fireEvent.press(getByText(Unit.metric));
    fireEvent.changeText(getAllByPlaceholderText('0')[1], kilograms);
    fireEvent.press(getByText(Unit.imperial));
    // Assert
    expect(getByDisplayValue(kgToLb(kilograms).toString())).toBeTruthy();
  });

  it('saves data to local storage', () => {
    // Arrange
    const feet = 5;
    const {getAllByPlaceholderText, getByText} = render(component);
    // Act
    fireEvent.changeText(getAllByPlaceholderText('0')[0], feet);
    fireEvent.press(getByText('Save'));
    // Assert
    expect(AsyncStorage.setItem).toHaveBeenCalledWith(
      STORAGE_KEY,
      JSON.stringify({
        ...initialScreenState,
        feet,
        meters: inToM(feet * 12),
      }),
    );
  });

  it('clears data from screen and local storage', async () => {
    // Arrange
    const pounds = 100;
    const {getAllByPlaceholderText, getByText} = render(component);
    // Act, Assert
    fireEvent.changeText(getAllByPlaceholderText('0')[2], pounds);
    await act(async () => await fireEvent.press(getByText('Save')));
    await act(async () => await fireEvent.press(getByText('Clear')));
    // Assert
    expect(AsyncStorage.removeItem).toHaveBeenCalledWith(STORAGE_KEY);
  });
});
