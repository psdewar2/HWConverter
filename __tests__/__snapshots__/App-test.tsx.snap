// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <RCTScrollView
    contentInsetAdjustmentBehavior="automatic"
  >
    <View>
      <View
        style={
          Object {
            "marginTop": 32,
            "paddingHorizontal": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
              "fontWeight": "600",
            }
          }
        >
          HW Converter
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-evenly",
              "marginVertical": 16,
            }
          }
        >
          <View
            accessibilityLabel="Imperial button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 4,
                  "elevation": 3,
                  "justifyContent": "center",
                  "paddingHorizontal": 32,
                  "paddingVertical": 12,
                },
                Object {
                  "backgroundColor": "lightgray",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "fontWeight": "600",
                }
              }
            >
              imperial
            </Text>
          </View>
          <View
            accessibilityLabel="Metric button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 4,
                  "elevation": 3,
                  "justifyContent": "center",
                  "paddingHorizontal": 32,
                  "paddingVertical": 12,
                },
                Object {
                  "backgroundColor": "white",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "fontWeight": "200",
                }
              }
            >
              metric
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "paddingHorizontal": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 32,
              "fontWeight": "600",
            }
          }
        >
          Measurements
        </Text>
        <View
          style={
            Object {
              "marginVertical": 16,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "200",
              }
            }
          >
            Height
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "baseline",
                  "flexDirection": "row",
                  "flexShrink": 1,
                  "justifyContent": "center",
                  "margin": 8,
                }
              }
            >
              <TextInput
                keyboardType="decimal-pad"
                maxLength={10}
                onChangeText={[Function]}
                placeholder="0"
                style={
                  Object {
                    "fontSize": 64,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "fontSize": 48,
                    "fontWeight": "300",
                    "marginLeft": 8,
                  }
                }
              >
                ft
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "baseline",
                  "flexDirection": "row",
                  "flexShrink": 1,
                  "justifyContent": "center",
                  "margin": 8,
                }
              }
            >
              <TextInput
                keyboardType="decimal-pad"
                maxLength={10}
                onChangeText={[Function]}
                placeholder="0"
                style={
                  Object {
                    "fontSize": 64,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "fontSize": 48,
                    "fontWeight": "300",
                    "marginLeft": 8,
                  }
                }
              >
                in
              </Text>
            </View>
          </View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "200",
              }
            }
          >
            Weight
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "baseline",
                  "flexDirection": "row",
                  "flexShrink": 1,
                  "justifyContent": "center",
                  "margin": 8,
                }
              }
            >
              <TextInput
                keyboardType="decimal-pad"
                maxLength={10}
                onChangeText={[Function]}
                placeholder="0"
                style={
                  Object {
                    "fontSize": 64,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "fontSize": 48,
                    "fontWeight": "300",
                    "marginLeft": 8,
                  }
                }
              >
                lbs
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-evenly",
              "marginVertical": 16,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Save
              </Text>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Clear
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
